File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.js

// class UnionFind{
//   constructor(x){
//     this.parents = Array(x).fill(0).map((_,index) => index); 
//     this.size = Array(x).fill(1);
//   }

//   find(x){
//     if(this.parents[x] === x){
//       return x;
//     }
//     return this.parents[x] = find(this.parents[x]);
//   }

//   union(x,y){
//     const rootX = this.find(x);
//     const rootY = this.find(y);

//     if(rootX !== rootY){
//       if(this.size[rootX] > this.size[rootY]){
//         this.parents[rootY] = rootX;
//         this.size[rootX] += this.size[rootY];
//       }else{
//         this.parents[rootX] = rootY;
//         this.size[rootY] += this.size[rootX];
//       }
//     }
//   }

//   getSize(x){
//     return this.size[this.find(x)];
//   }
// }

// function solution(land) {
//   var answer = 0;
//   const height = land.length;
//   const width = land[0].length;

//   const dx = [1,-1,0,0];
//   const dy = [0,0,1,-1];

//   var uf = new UnionFind(height * width);

//   for(let i = 0; i<height; i++){
//     for(let j = 0; j<width; j++){
//       if(land[i][j] === 1){
//         for(let k = 0; k<4; k++){
//           let nx = i + dx[k];
//           let ny = j + dy[k];
//           if(nx >= 0 && ny >= 0 && nx < height && ny < width && land[nx][ny] === 1){
//             uf.union(i * width + j , nx * width + ny);
//           }
//         }
//       }
//     }
//   }

//   var oil = Array(width).fill(0);
//   for(let j = 0; j<width; j++){
//     var visited = Array(width * height).fill(0);
//     for(let i = 0; i<height; i++){
//       if(land[i][j] === 1){
//         var root = uf.find(i*width + j);
//         if(!visited[root]){
//           oil[j] += uf.getSize(root);
//           visited[root] = true;
//         }
//       }
//     }
//   }
  
//   return Math.max(...oil);
// }






function solution(land) {
  
  const width = land[0].length;
  const height = land.length;
  const visited = Array.from({ length: height }, () => Array(width).fill(false));
  const dx = [1,-1,0,0];
  const dy = [0,0,1,-1];

  var rowVal = Array(width).fill(0);

  function bfs (x, y){
    const q = [[x,y]];
    const visitedRow = new Set();
    visited[x][y] = true;
    let cnt = 0;

    while(q.length > 0){
      const [cx, cy] = q.shift();
      cnt++;
      visitedRow.add(cy);

      for(let i = 0; i<4; i++){
        const nx = cx + dx[i];
        const ny = cy + dy[i];
        if(nx >= 0 && ny >= 0 && nx < height && ny < width && land[nx][ny] === 1 && !visited[nx][ny]){
          q.push([nx, ny]);
          visited[nx][ny] = true;
        }
      }
    }
    for(const s of visitedRow){
      rowVal[s] += cnt;
    }
  }

  for(let j = 0 ; j<width; j++){
    for(let i = 0; i<height; i++){
      if(land[i][j] === 1 && !visited[i][j]){
        bfs(i,j);
      }
    }
  }




  return Math.max(...rowVal);
}

File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.java
import java.util.HashSet;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Set;

class UnionFind {
  private int[] parent;
  private int[] size;

  public UnionFind(int n) {
    parent = new int[n];
    size = new int[n];

    for (int i = 0; i < n; i++) {
      parent[i] = i; // 초기화시 자신의 부모는 자신
      size[i] = 1;  // 초기화시 각 node 의 크기는 1
    }
  }

  /*
  필요한 연산
  1. root 찾기
  2. union
  3. size 찾기
   */
  public int find(int x) {
    if (parent[x] != x) {
      parent[x] = find(parent[x]);
    }
    return parent[x];
  }

  public void union(int x, int y) {
    int rootX = find(x);
    int rootY = find(y);

    if (rootX == rootY) return; // 이미 같은 집합

    if (size[rootX] > size[rootY]) { // 크기가 더 큰쪽을 부모로
      parent[rootY] = rootX;
      size[rootX] += size[rootY];
    } else {
      parent[rootX] = rootY;
      size[rootY] += size[rootX];
    }
  }

  public int getSize(int x) {
    return size[find(x)];
  }
}

class Solution {
  int[] dx = {1, -1, 0, 0};
  int[] dy = {0, 0, 1, -1};
  boolean[][] visited;
  int n, m;

  public int solution(int[][] land) {
    int answer = 0;

    n = land.length;
    m = land[0].length;
    visited = new boolean[n][m];
    UnionFind uf = new UnionFind(n * m);

    for (int i = 0; i < n; i++) {
      for (int j = 0; j < m; j++) {
        if (land[i][j] == 1) {
          for(int k = 0; k<4; k++){
            int ni = i + dx[k];
            int nj = j + dy[k];
            if((ni >= 0 && nj >=0 && ni < n && nj < m) && land[ni][nj] == 1){
              uf.union(i * m + j, ni * m + nj);
            }
          }
        }
      }
    }

    Set<Integer> roots =new HashSet<>();

    for(int i = 0; i<m; i++){
      for(int j = 0; j<n; j++){
        if(land[j][i] == 1){
          roots.add(uf.find(j *m + i));
        }
      }
      
      int sum = 0;
      for(Integer r : roots){
        sum += uf.getSize(r);
      }
      
      answer = Math.max(answer, sum);
      roots.clear();
    }

    return answer;
  }
}